File: 程序员面试金典/4.6后继者/inorderSuccessor.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private Integer last = null;

    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        if (root == null)
            return null;
        TreeNode suc = inorderSuccessor(root.left, p);
        if (suc != null)
            return suc;
        if (last != null && last == p.val)
            return root;
        last = root.val;
        return inorderSuccessor(root.right, p);
    }
}

File: 程序员面试金典/4.6/inorderSuccessor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:

    def __init__(self):
        self.pre = None
    
    def inorderSuccessor(self, root: TreeNode, p: TreeNode) -> TreeNode:
        if root is None:
            return None
        suc = self.inorderSuccessor(root.left, p)
        if suc is not None:
            return suc
        if self.pre == p.val:
            return root
        self.pre = root.val
        return self.inorderSuccessor(root.right, p)
